vmm/src/interrupt.rs

```diff
@@ -23,6 +23,10 @@ struct InterruptRoute {
     gsi: u32,
     irq_fd: EventFd,
     registered: AtomicBool,
+    #[cfg(target_arch = "x86_64")]
+    resample_fd: Mutex<Option<EventFd>>,
```

Comment:
I think the related code can be simplified if you use `resample_fd: EventFd`. 

You just need to initialize it from the constructor function e.g. `InterruptRoute::new()`, and use `resample_enabled` to decide whether a `unmask_notifier()` will return `None` or `Some`.

Reply:
Currently, only VFIO devices will use resamplefd. If we create the resamplefd in the new() function, then all the IOAPIC and MSI/MSIX interrupts will have a resamplefd which is unnecessary. Therefore, I added a new interface called enable_resamplefd, which only creates the resamplefd when it is needed.

Reply:
Actually, my confusion was from using both `Mutex` or the `AtomicBool` at the same time.

IIUC, we only need to avoid racing for initializing the `resample_fd` (which is the only "write") and all following accesses are "read"(e.g. `clone()`). So I think either one of the following would work, right?
```
    #[cfg(target_arch = "x86_64")]
    resample_fd: Mutex<Option<EventFd>>,
```

or 
```
    #[cfg(target_arch = "x86_64")]
    resample_fd: Option<EventFd>,
    #[cfg(target_arch = "x86_64")]
    resample_enabled: AtomicBool,
```

Reply:
Thank you very much for your correction, I think you are right.

Reply:
> Actually, my confusion was from using both `Mutex` or the `AtomicBool` at the same time.
> 
> IIUC, we only need to avoid racing for initializing the `resample_fd` (which is the only "write") and all following accesses are "read"(e.g. `clone()`). So I think either one of the following would work, right?
> 
> ```
>     #[cfg(target_arch = "x86_64")]
>     resample_fd: Mutex<Option<EventFd>>,
> ```
> 
@rbradford I have used a Mutex to serialize access and clean the AtomicBool variable in the latest patch.

> or
> 
> ```
>     #[cfg(target_arch = "x86_64")]
>     resample_fd: Option<EventFd>,
>     #[cfg(target_arch = "x86_64")]
>     resample_enabled: AtomicBool,
> ```